pkg/wasm/convert.go

```diff
@@ -0,0 +1,153 @@
+	newExtensionConfig = resource
+	sendNack = false
+	if err := ptypes.UnmarshalAny(resource, ec); err != nil {
+		wasmLog.Debugf("failed to unmarshal extension config resource: %v", err)
+		return
+	}
+
+	// Currently Wasm filter can only be configured using typed struct via EnvoyFilter.
+	wasmLog.Debugf("original extension config resource %+v", ec)
+	if ec.GetTypedConfig() == nil && ec.GetTypedConfig().TypeUrl != typedStructType {
+		wasmLog.Debugf("cannot find typed struct in %+v", ec)
+		return
+	}
+	wasmStruct := &udpa.TypedStruct{}
+	wasmTypedConfig := ec.GetTypedConfig()
+	if err := ptypes.UnmarshalAny(wasmTypedConfig, wasmStruct); err != nil {
+		wasmLog.Debugf("failed to unmarshal typed config for wasm filter: %v", err)
+		return
+	}
+
+	if wasmStruct.TypeUrl != wasmHTTPFilterType {
+		wasmLog.Debugf("typed extension config %+v does not contain wasm http filter", wasmStruct)
+		return
+	}
+
+	wasmHTTPFilterConfig := &wasm.Wasm{}
+	if err := conversion.StructToMessage(wasmStruct.Value, wasmHTTPFilterConfig); err != nil {
+		wasmLog.Debugf("failed to convert extension config struct %+v to Wasm HTTP filter", wasmStruct)
+		return
+	}
+
+	if wasmHTTPFilterConfig.Config == nil ||
+		wasmHTTPFilterConfig.Config.GetVmConfig() == nil ||
+		wasmHTTPFilterConfig.Config.GetVmConfig().GetCode() == nil ||
+		wasmHTTPFilterConfig.Config.GetVmConfig().GetCode().GetRemote() == nil {
+		wasmLog.Debugf("no remote load found in Wasm HTTP filter %+v", wasmHTTPFilterConfig)
+		return
+	}
+
+	// Wasm plugin configuration has remote load. From this point, any failure should result as a Nack,
+	// unless the plugin is marked as fail open.
+	failOpen := wasmHTTPFilterConfig.Config.GetFailOpen()
+	sendNack = !failOpen
+
+	vm := wasmHTTPFilterConfig.Config.GetVmConfig()
+	remote := vm.GetCode().GetRemote()
+	httpURI := remote.GetHttpUri()
+	if httpURI == nil {
+		wasmLog.Errorf("wasm remote fetch %+v does not have httpUri specified", remote)
+		return
+	}
+	timeout := time.Duration(0)
+	if remote.GetHttpUri().Timeout != nil {
+		timeout = remote.GetHttpUri().Timeout.AsDuration()
+	}
+	f, err := cache.Get(httpURI.GetUri(), remote.GetSha256(), timeout)
+	if err != nil {
+		wasmLog.Errorf("cannot fetch Wasm module %v: %v", remote.GetHttpUri().GetUri(), err)
+		return
+	}
+
+	// Rewrite remote fetch to local file.
+	vm.Code = &core.AsyncDataSource{
+		Specifier: &core.AsyncDataSource_Local{
+			Local: &core.DataSource{
+				Specifier: &core.DataSource_Filename{
+					Filename: f,
+				},
+			},
+		},
+	}
+
+	wasmTypedConfig, err = ptypes.MarshalAny(wasmHTTPFilterConfig)
+	if err != nil {
+		wasmLog.Errorf("failed to marshal new wasm HTTP filter %+v to protobuf Any: %v", wasmHTTPFilterConfig, err)
+		return
+	}
+	ec.TypedConfig = wasmTypedConfig
+	wasmLog.Debugf("new extension config resource %+v", ec)
+
+	newExtensionConfig, err = ptypes.MarshalAny(ec)
+	if err != nil {
+		wasmLog.Errorf("failed to marshal new extension config resource: %v", err)
+		return
```

Comment:
if this fails, but its fail open, we will `return nil, false`. We do happen to filter out the `nil` later, buti s it expected?

Reply:
Good catch, the new extension should assign after all steps. Updated.